sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproxCountDistinctForIntervals.scala

```diff
@@ -0,0 +1,248 @@
+      number of distinct values (ndv) for intervals [endpoint_1, endpoint_2],
+      (endpoint_2, endpoint_3], ... (endpoint_N-1, endpoint_N].
+
+    _FUNC_(col, array(endpoint_1, endpoint_2, ... endpoint_N), relativeSD=0.05) - Returns
+      the approximate number of distinct values (ndv) for intervals with relativeSD, the maximum
+      estimation error allowed in the HyperLogLogPlusPlus algorithm.
+  """,
+  extended = """
+    Examples:
+      > SELECT approx_count_distinct_for_intervals(10.0, array(5, 15, 25), 0.01);
+       [1, 0]
+  """)
+case class ApproxCountDistinctForIntervals(
+    child: Expression,
+    endpointsExpression: Expression,
+    relativeSD: Double = 0.05,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+  extends ImperativeAggregate with ExpectsInputTypes {
+
+  def this(child: Expression, endpointsExpression: Expression) = {
+    this(
+      child = child,
+      endpointsExpression = endpointsExpression,
+      relativeSD = 0.05,
+      mutableAggBufferOffset = 0,
+      inputAggBufferOffset = 0)
+  }
+
+  def this(child: Expression, endpointsExpression: Expression, relativeSD: Expression) = {
+    this(
+      child = child,
+      endpointsExpression = endpointsExpression,
+      relativeSD = HyperLogLogPlusPlus.validateDoubleLiteral(relativeSD),
+      mutableAggBufferOffset = 0,
+      inputAggBufferOffset = 0)
+  }
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    Seq(TypeCollection(NumericType, TimestampType, DateType), ArrayType)
+  }
+
+  // Mark as lazy so that endpointsExpression is not evaluated during tree transformation.
+  lazy val endpoints: Array[Double] = {
+    val doubleArray = (endpointsExpression.dataType, endpointsExpression.eval()) match {
+      case (ArrayType(baseType: NumericType, _), arrayData: ArrayData) =>
+        val numericArray = arrayData.toObjectArray(baseType)
+        numericArray.map { x =>
+          baseType.numeric.toDouble(x.asInstanceOf[baseType.InternalType])
+        }
+    }
+    util.Arrays.sort(doubleArray)
+    doubleArray
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      defaultCheck
+    } else if (!endpointsExpression.foldable) {
+      TypeCheckFailure("The intervals provided must be constant literals")
+    } else if (endpoints.length < 2) {
+      TypeCheckFailure("The number of endpoints must be >= 2 to construct intervals")
+    } else {
+      TypeCheckSuccess
+    }
+  }
+
+  // N endpoints construct N-1 intervals, creating a HLLPP for each interval
+  private lazy val hllppArray = {
+    val array = new Array[HyperLogLogPlusPlusHelper](endpoints.length - 1)
+    for (i <- array.indices) {
+      array(i) = new HyperLogLogPlusPlusHelper(relativeSD)
+    }
+    array
+  }
+
+  private lazy val numWordsPerHllpp = hllppArray.head.numWords
```

Comment:
do all hllpphelper have same value of `numWords`?

Reply:
yes, `numWords` is determined by parameter `relativeSD`, which is the same for all HLLPPHelper.